### PR TITLE
Adding plots vs TPC occupancy

### PR DESCRIPTION
**Pull request by chiarazampolli**

### CODE REVIEW

**Comment by chiarazampolli** [comment]
Do not review yet, I need to see if it compiles with QC.

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 

If it compiles fine, this would be ready for review.

Chiara

**Comment by chiarazampolli** [comment]
An example from 3 TFs from PbPb (before fixing the Z axis title which is in the last commit):

Screenshot 2024-10-04 at 10 46 17

And from pp:
Screenshot 2024-10-04 at 10 48 44

I am not sure why the multiplicity shows an underflow (in pp; in PbPb there are 15)... The limit is 0, so I would assume there is no underflow.

**Comment by chiarazampolli** [comment]
As for the underflow, can it be some artifact? If I take the projectionZ itself, I see that the bin with 15 entries is not the underflow one:

{code}
root [3] .x plot.C
ibinZ = 0 low edge = -1e+06 content = 0
ibinZ = 1 low edge = 0 content = 15
ibinZ = 2 low edge = 1e+06 content = 466
ibinZ = 3 low edge = 2e+06 content = 1148
ibinZ = 4 low edge = 3e+06 content = 1139
ibinZ = 5 low edge = 4e+06 content = 951
ibinZ = 6 low edge = 5e+06 content = 2816
{code}

The macro I used to plot and then print the bins is:

{code}
void plot(bool isPbPb = true) {

  TFile* f = new TFile("ITSTPCmatched_allTracks_enableK0.root");
  TString plotName = "mK0MassVsPtVsOccPbPb";
  if (isPbPb == false) {
    plotName = "mK0MassVsPtVsOccpp";
  }
  TH3F* h1 = (TH3F*)f->Get(Form("%s", plotName.Data()));
  TH1D* hx = (TH1D*)h1->ProjectionX();
  TH1D* hy = (TH1D*)h1->ProjectionY();
  TH1D* hz = (TH1D*)h1->ProjectionZ();
  gStyle->SetOptStat(1111111);

  TCanvas* c = new TCanvas("c", "c", 50, 50, 1150, 1150);
  c->Divide(2, 2);
  c->cd(1);
  h1->Draw();
  c->cd(2);
  hx->Draw();
  c->cd(3);
  hy->Draw();
  c->cd(4);
  hz->Draw();

  for (int ibinX = 1; ibinX <= h1->GetNbinsX(); ++ibinX) {
    for (int ibinY = 1; ibinY <= h1->GetNbinsY(); ++ibinY) {
      if (h1->GetBinContent(ibinX, ibinY, 0) != 0) {
	std::cout << "ibinX = " << ibinX << " ibiny = " << ibinY << " content with binZ=0 = " << h1->GetBinContent(ibinX, ibinY, 0) << std::endl;
      }
    }
  }
  for (int ibinZ = 0; ibinZ <= hz->GetNbinsX(); ++ibinZ) {
    std::cout << "ibinZ = " << ibinZ << " low edge = " << hz->GetXaxis()->GetBinLowEdge(ibinZ) << " content = " << hz->GetBinContent(ibinZ) << std::endl;
  }
    //  c->SaveAs("K0s.png");
    //c->SaveAs("K0s.pdf");
  return;
}
{code}

**Comment by chiarazampolli** [comment]
The observation about underflow is simply caused by the fact that I have changed to start from occupancy 0 but probably did not re-run all tests (I am doing 4, pp, PbPb, with and without QC each). Re-running everything, all is good, and when the occ axis starts from 0, there are no underflows.

**pp:**
Screenshot 2024-10-04 at 14 35 46

with:
ibinZ = 0 low edge = -166667 content = 0
ibinZ = 1 low edge = 0 content = 1
ibinZ = 2 low edge = 166667 content = 2
ibinZ = 3 low edge = 333333 content = 37
ibinZ = 4 low edge = 500000 content = 164
ibinZ = 5 low edge = 666667 content = 35
ibinZ = 6 low edge = 833333 content = 2


**PbPb:**
Screenshot 2024-10-04 at 14 37 26

with:

ibinZ = 0 low edge = -1e+06 content = 0
ibinZ = 1 low edge = 0 content = 15
ibinZ = 2 low edge = 1e+06 content = 466
ibinZ = 3 low edge = 2e+06 content = 1148
ibinZ = 4 low edge = 3e+06 content = 1139
ibinZ = 5 low edge = 4e+06 content = 951
ibinZ = 6 low edge = 5e+06 content = 2816

**Comment by chiarazampolli** [comment]
Moving to WIP. The current binning in PbPb, with the reco of a full CTF at ~ 26 kHz shows 1.7% of overflow in TPC occupancy. I am testing with a wider range.

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 

For me this can be merged. 
You will see that there are 2 histograms 3D, one for pp and one for PbPb, where the binning in TPC occupancy is different. Within the QC framework, one could decide which one to use, but it needs to be checked and implemented in the async merging stage. 
Without the QC framework, one could then use the configParam. 
For now, we have two histograms, one will be empty.

Chiara

**Comment by chiarazampolli** [comment]
With a second thought, maybe it is better to add the cuts on DCA and cosPA before merging...

**Comment by shahor02** [comment]
OK, I would make the cuts the parameters of the json. 
To avoid defining 2 histos: cannot we use different jsons (passing the binning) depending on the `$BEAMTYPE` in the qc-workflow.sh ?

**Comment by chiarazampolli** [comment]
Yes, sure, the cuts will be a parameter. 
And yes, 2 jsons were the first thought I had, then I found this:
https://github.com/AliceO2Group/QualityControl/blob/master/doc/Advanced.md#definition-and-access-of-user-defined-configuration-extendedtaskparameters
which in online would work very well, but for async we need to see for the merging, it seems. 
I think that maintaining 2 json is very error prone, this is why I would not do it unless this extra histogram is really a problem.

**Comment by chiarazampolli** [comment]
With the cut on DCA and cosPA, 1 full CTF from zzo (PbPb): statistics decrease from ~170k to ~3k candidates: 

Screenshot 2024-10-10 at 13 42 50

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 
Can you review?
Thanks!
Chiara

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 

As discussed, this PR could now be reviewed, and once ok, merged. The K0s are anyway not yet enabled. To do that we should clarify the issue with the inputs, as discussed via email. 

Chiara

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 
Please note that I will be away next week, so in case you ask for modifications, I will implement them only when I am back.
Chiara

File: Detectors/GLOQC/src/MatchITSTPCQC.cxx [hunk 1]
```diff
@@ -375,9 +385,50 @@ bool MatchITSTPCQC::init()
...
+
+  const Int_t nbinsMultK0PbPb = mNBinsTPCOccPbPb;
+  Double_t* zbinsMultK0PbPb = new Double_t[nbinsMultK0PbPb + 1];
+  Double_t zminMultK0PbPb = mMinTPCOccPbPb;
+  Double_t zmaxMultK0PbPb = mMaxTPCOccPbPb;
+  Double_t dzMultK0PbPb = (zmaxMultK0PbPb - zminMultK0PbPb) / nbinsMultK0PbPb;
+  for (int i = 0; i <= nbinsMultK0PbPb; i++) {
+    zbinsMultK0PbPb[i] = zminMultK0PbPb + i * dzMultK0PbPb;
+  }
+
```
Comment (shahor02) [inline Detectors/GLOQC/src/MatchITSTPCQC.cxx:426]: one should `delete[] ` axes binning arrays booked with `new`.

**Review by shahor02 — CHANGES_REQUESTED** [review]
Hi @chiarazampolli 
just one comment below.

File: Detectors/GLOQC/src/MatchITSTPCQC.cxx (hunk 1 above)
Comment (chiarazampolli) [reply Detectors/GLOQC/src/MatchITSTPCQC.cxx:426]: Done.

**Review by shahor02 — APPROVED** [review]

**Pull request merged**